File: android/src/main/kotlin/com/pycampers/flutter_cognito_plugin/Cognito.kt
package com.pycampers.flutter_cognito_plugin

import android.app.Activity
import android.content.Context
import com.amazonaws.auth.CognitoCachingCredentialsProvider
import com.amazonaws.mobile.client.AWSMobileClient
import com.amazonaws.mobile.client.Callback
import com.amazonaws.mobile.client.HostedUIOptions
import com.amazonaws.mobile.client.SignInUIOptions
import com.amazonaws.mobile.client.SignOutOptions
import com.amazonaws.mobile.client.UserStateDetails
import com.amazonaws.mobile.client.results.ForgotPasswordResult
import com.amazonaws.mobile.client.results.SignInResult
import com.amazonaws.mobile.client.results.SignUpResult
import com.amazonaws.mobile.client.results.Tokens
import com.amazonaws.mobile.client.results.UserCodeDeliveryDetails
import com.pycampers.plugin_scaffold.sendThrowable
import com.pycampers.plugin_scaffold.trySend
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

class Cognito(val context: Context) {
    var activity: Activity? = null
    val awsClient = AWSMobileClient.getInstance()!!

    fun initialize(call: MethodCall, result: Result) {
        awsClient.initialize(context, object : Callback<UserStateDetails> {
            override fun onResult(u: UserStateDetails) {
                trySend(result) { dumpUserState(u) }
            }

            override fun onError(e: Exception) {
                sendThrowable(result, e)
            }
        })
    }

    fun signUp(call: MethodCall, result: Result) {
        val username = call.argument<String>("username")
        val password = call.argument<String>("password")
        val userAttributes = call.argument<Map<String, String>>("userAttributes")

        awsClient.signUp(username, password, userAttributes, null, object : Callback<SignUpResult> {
            override fun onResult(s: SignUpResult) = trySend(result) { dumpSignUpResult(s) }
            override fun onError(e: Exception) = sendThrowable(result, e)
        })
    }

    fun confirmSignUp(call: MethodCall, result: Result) {
        val username = call.argument<String>("username")
        val confirmationCode = call.argument<String>("confirmationCode")

        awsClient.confirmSignUp(username, confirmationCode, object : Callback<SignUpResult> {
            override fun onResult(s: SignUpResult) = trySend(result) { dumpSignUpResult(s) }
            override fun onError(e: Exception) = sendThrowable(result, e)
        })
    }

    fun resendSignUp(call: MethodCall, result: Result) {
        val username = call.argument<String>("username")

        awsClient.resendSignUp(username, object : Callback<SignUpResult> {
            override fun onResult(s: SignUpResult) = trySend(result) { dumpSignUpResult(s) }
            override fun onError(e: Exception) = sendThrowable(result, e)
        })
    }

    fun signIn(call: MethodCall, result: Result) {
        val username = call.argument<String>("username")
        val password = call.argument<String>("password")

        awsClient.signIn(username, password, null, object : Callback<SignInResult> {
            override fun onResult(s: SignInResult) = trySend(result) { dumpSignInResult(s) }
            override fun onError(e: Exception) = sendThrowable(result, e)
        })
    }

    fun confirmSignIn(call: MethodCall, result: Result) {
        val confirmationCode = call.argument<String>("confirmationCode")

        awsClient.confirmSignIn(confirmationCode, object : Callback<SignInResult> {
            override fun onResult(s: SignInResult) = trySend(result) { dumpSignInResult(s) }
            override fun onError(e: Exception) = sendThrowable(result, e)
        })
    }

    fun changePassword(call: MethodCall, result: Result) {
        val oldPassword = call.argument<String>("oldPassword")
        val newPassword = call.argument<String>("newPassword")

        awsClient.changePassword(oldPassword, newPassword, object : Callback<Void>  {
            override fun onResult(u: Void?) = trySend(result)
            override fun onError(e: Exception) = sendThrowable(result, e)
        })
    }

    fun forgotPassword(call: MethodCall, result: Result) {
        val username = call.argument<String>("username")

        awsClient.forgotPassword(username, object : Callback<ForgotPasswordResult> {
            override fun onResult(f: ForgotPasswordResult) {
                trySend(result) { dumpForgotPasswordResult(f) }
            }

            override fun onError(e: Exception) = sendThrowable(result, e)
        })
    }

    fun confirmForgotPassword(call: MethodCall, result: Result) {
        val newPassword = call.argument<String>("newPassword")
        val confirmationCode = call.argument<String>("confirmationCode")

        awsClient.confirmForgotPassword(
            newPassword,
            confirmationCode,
            object : Callback<ForgotPasswordResult> {
                override fun onResult(f: ForgotPasswordResult) =
                    trySend(result) { dumpForgotPasswordResult(f) }

                override fun onError(e: Exception) = sendThrowable(result, e)
            })
    }

    fun getUserAttributes(call: MethodCall, result: Result) {
        awsClient.getUserAttributes(object : Callback<Map<String, String>> {
            override fun onResult(attrs: Map<String, String>?) = trySend(result) { attrs }
            override fun onError(e: Exception) = sendThrowable(result, e)
        })
    }

    fun updateUserAttributes(call: MethodCall, result: Result) {
        val userAttributes = call.argument<Map<String, String>>("userAttributes")!!

        awsClient.updateUserAttributes(
            userAttributes,
            object : Callback<List<UserCodeDeliveryDetails>> {
                override fun onResult(u: List<UserCodeDeliveryDetails>) {
                    trySend(result) {
                        u.map { dumpUserCodeDeliveryDetails(it) }
                    }
                }

                override fun onError(e: Exception) = sendThrowable(result, e)
            }
        )
    }

    fun confirmUpdateUserAttribute(call: MethodCall, result: Result) {
        val attributeName = call.argument<String>("attributeName")!!
        val confirmationCode = call.argument<String>("confirmationCode")!!

        awsClient.confirmUpdateUserAttribute(
            attributeName,
            confirmationCode,
            object : Callback<Void> {
                override fun onResult(u: Void) = trySend(result) { u }
                override fun onError(e: Exception) = sendThrowable(result, e)
            })
    }

    fun signOut(call: MethodCall, result: Result) {
        val invalidateTokens = call.argument<Boolean>("invalidateTokens")!!
        val signOutGlobally = call.argument<Boolean>("signOutGlobally")!!

        val options = SignOutOptions
            .builder()
            .invalidateTokens(invalidateTokens)
            .signOutGlobally(signOutGlobally)
            .build()

        awsClient.signOut(
            options,
            object : Callback<Void> {
                override fun onResult(v: Void?) = trySend(result)
                override fun onError(e: Exception) = sendThrowable(result, e)
            }
        )
    }

    fun getUsername(call: MethodCall, result: Result) {
        result.success(awsClient.username)
    }

    fun isSignedIn(call: MethodCall, result: Result) {
        result.success(awsClient.isSignedIn)
    }

    fun getIdentityId(call: MethodCall, result: Result) {
        result.success(awsClient.identityId)
    }

    fun currentUserState(call: MethodCall, result: Result) {
        result.success(awsClient.currentUserState()?.userState?.ordinal)
    }

    fun getTokens(call: MethodCall, result: Result) {
        awsClient.getTokens(object : Callback<Tokens> {
            override fun onResult(t: Tokens?) = trySend(result) { t?.let { dumpTokens(it) } ?: t }
            override fun onError(e: Exception) = sendThrowable(result, e)
        })
    }

    fun getCredentials(call: MethodCall, result: Result) {
        result.success(
            dumpCredentials(
                CognitoCachingCredentialsProvider(
                    context,
                    awsClient.configuration
                )
            )
        )
    }

    fun federatedSignIn(call: MethodCall, result: Result) {
        val providerName = call.argument<String>("providerName")
        val token = call.argument<String>("token")

        awsClient.federatedSignIn(providerName, token, object : Callback<UserStateDetails> {
            override fun onResult(u: UserStateDetails) = trySend(result) { dumpUserState(u) }
            override fun onError(e: Exception) = sendThrowable(result, e)
        })
    }

    fun showSignIn(call: MethodCall, result: Result) {
        if (activity == null) {
            result.error(
                "ActivityNotAvailable",
                "This method cannot be called without an Activity.\n" +
                    "Did you forget to replace `FlutterActivity` by `CognitoPluginActivity` in your app's    MainActivity.kt?",
                null
            )
            return
        }

        val scopes = call.argument<List<String>>("scopes")!!
        val identityProvider = call.argument<String>("identityProvider")!!

        val hostedUIOptions = HostedUIOptions.builder()
            .scopes(*scopes.toTypedArray())
            .identityProvider(identityProvider)
            .build()
        val signInUIOptions = SignInUIOptions.builder()
            .hostedUIOptions(hostedUIOptions)
            .build()

        AWSMobileClient.getInstance().showSignIn(
            activity,
            signInUIOptions,
            object : Callback<UserStateDetails> {
                override fun onResult(u: UserStateDetails) = trySend(result) { dumpUserState(u) }
                override fun onError(e: Exception) = sendThrowable(result, e)
            }
        )
    }
}

File: ios/Classes/Cognito.swift
import AWSMobileClient
import Flutter
import plugin_scaffold

typealias CompletionCallback<T> = (T?, Error?) -> Void

class Cognito {
    var awsClient: AWSMobileClient?
    var userStateCallback: UserStateChangeCallback?
    private lazy var notInitializedError = FlutterError(
        code: "NotInitialized",
        message: "Call the initialize method first.",
        details: nil
    )

    func createErrorCallback(_ result: @escaping FlutterResult) -> (Error?) -> Void {
        return { error in
            if let error = error {
                trySendError(result, error)
            } else {
                trySend(result) { nil }
            }
        }
    }

    func createCallback<T>(_ result: @escaping FlutterResult, _ serializer: ((T) -> Any)? = nil) -> (T?, Error?) -> Void {
        return { obj, error in
            if let obj = obj {
                trySend(result) {
                    let value = serializer?(obj) ?? obj
                    return value
                }
            } else if let error = error {
                trySendError(result, error)
            }
        }
    }

    func initialize(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any?]
        if let configuration = args["configuration"] as? String,
           let url = Bundle.main.url(forResource: "awsconfiguration", withExtension: "json"),
           let data = try? Data(contentsOf: url),
           let obj = try? JSONSerialization.jsonObject(with: data) as? [String:Any] {
            var config:[String: Any] = [:]
            obj.forEach { (key, value) in
                if var value = value as? [String:Any],
                   let newval = value[configuration] {
                    value["Default"] = newval
                    value[configuration] = nil
                    config[key] = value
                } else {
                    config[key] = value
                }
            }
            self.awsClient = AWSMobileClient(configuration: config)
        } else {
            self.awsClient = AWSMobileClient.default()
        }
        if let callback = userStateCallback {
            let obj = "test" as NSString
            self.awsClient?.removeUserStateListener(obj)
            self.awsClient?.addUserStateListener(obj, callback)
        }
        self.awsClient?.initialize(self.createCallback(result, dumpUserState))
    }

    func signUp(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let username = args["username"] as! String
        let password = args["password"] as! String
        let attrs = args["userAttributes"] as! [String: String]

        awsClient.signUp(
            username: username,
            password: password,
            userAttributes: attrs,
            validationData: [:],
            completionHandler: createCallback(result, dumpSignUpResult)
        )
    }

    func confirmSignUp(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let username = args["username"] as! String
        let confirmationCode = args["confirmationCode"] as! String

        awsClient.confirmSignUp(
            username: username,
            confirmationCode: confirmationCode,
            completionHandler: createCallback(result, dumpSignUpResult)
        )
    }

    func resendSignUp(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let username = args["username"] as! String

        awsClient.resendSignUpCode(
            username: username,
            completionHandler: createCallback(result, dumpSignUpResult)
        )
    }

    func signIn(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let username = args["username"] as! String
        let password = args["password"] as! String

        awsClient.signIn(
            username: username,
            password: password,
            validationData: [:],
            completionHandler: createCallback(result, dumpSignInResult)
        )
    }

    func confirmSignIn(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let confirmationCode = args["confirmationCode"] as! String

        awsClient.confirmSignIn(
            challengeResponse: confirmationCode,
            completionHandler: createCallback(result, dumpSignInResult)
        )
    }

    func changePassword(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let oldPassword = args["oldPassword"] as! String
        let newPassword = args["newPassword"] as! String

        awsClient.changePassword(
            currentPassword: oldPassword,
            proposedPassword: newPassword,
            completionHandler: createErrorCallback(result)
        )
    }

    func forgotPassword(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let username = args["username"] as! String

        awsClient.forgotPassword(
            username: username,
            completionHandler: createCallback(result, dumpForgotPasswordResult)
        )
    }

    func confirmForgotPassword(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let username = args["username"] as! String
        let newPassword = args["newPassword"] as! String
        let confirmationCode = args["confirmationCode"] as! String

        awsClient.confirmForgotPassword(
            username: username,
            newPassword: newPassword,
            confirmationCode: confirmationCode,
            completionHandler: createCallback(result, dumpForgotPasswordResult)
        )
    }

    func updateUserAttributes(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let userAttributes = args["userAttributes"] as! [String: String]

        awsClient.updateUserAttributes(
            attributeMap: userAttributes,
            completionHandler: createCallback(result) {
                $0.map { dumpUserCodeDeliveryDetails($0) }
            }
        )
    }

    func getUserAttributes(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        awsClient.getUserAttributes(completionHandler: self.createCallback(result))
    }

    func confirmUpdateUserAttribute(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let attributeName = args["attributeName"] as! String
        let confirmationCode = args["confirmationCode"] as! String

        awsClient.confirmUpdateUserAttributes(
            attributeName: attributeName, code: confirmationCode,
            completionHandler: createErrorCallback(result)
        )
    }

    func signOut(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        awsClient.signOut()
        result(nil)
    }

    func getUsername(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        result(awsClient.username)
    }

    func isSignedIn(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        result(awsClient.isSignedIn)
    }

    func getIdentityId(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        result(awsClient.identityId)
    }

    func currentUserState(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        result(dumpUserState(awsClient.currentUserState))
    }

    func getTokens(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        awsClient.getTokens(self.createCallback(result, dumpTokens))
    }

    func getCredentials(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        awsClient.getAWSCredentials(self.createCallback(result, dumpCredentials))
    }

    func federatedSignIn(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let args = call.arguments as! [String: Any?]
        let providerName = args["providerName"] as! String
        let token = args["token"] as! String

        awsClient.federatedSignIn(
            providerName: providerName,
            token: token,
            completionHandler: createCallback(result, dumpUserState)
        )
    }

    func showSignIn(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let awsClient = self.awsClient else {
            result(notInitializedError)
            return
        }
        let navigationController = CognitoPluginAppDelegate.navigationController
        if navigationController == nil {
            let error = FlutterError(
                code: "UINavigationControllerNotFound",
                message: "This method cannot be called without access to a UINavigationController.\n" +
                    "Did you forget to replace `FlutterAppDelegate` with `CognitoPluginAppDelegate` in your app's AppDelegate.swift?",
                details: nil
            )
            result(error)
        } else {
            let args = call.arguments as! [String: Any?]
            let identityProvider = args["identityProvider"] as! String
            let scopes = args["scopes"] as! [String]

            // Optionally override the scopes based on the usecase.
            let hostedUIOptions = HostedUIOptions(scopes: scopes, identityProvider: identityProvider)

            // Present the Hosted UI sign in.
            awsClient.showSignIn(
                navigationController: navigationController!,
                hostedUIOptions: hostedUIOptions,
                createCallback(result, dumpUserState)
            )
        }
    }
}
